invenio_rdm_records/requests/access/models.py

```diff
@@ -0,0 +1,134 @@
+    """Expiration date."""
+
+    email = db.Column(db.String, nullable=False)
+    """E-mail address for which the access request was made."""
+
+    full_name = db.Column(db.String, nullable=False, default="")
+    """Full name of the requester."""
+
+    message = db.Column(db.Text, nullable=False, default="")
+    """Message associated with the request, explaining why access is needed."""
+
+    record_pid = db.Column(db.String, nullable=False)
+    """PID value of the record for which the request will be created."""
+
+    def delete(self):
+        """Delete this secret link."""
+        db.session.delete(self)
+
+    def to_dict(self):
+        """Dump the access request token to a dictionary."""
+        return {
+            "id": str(self.id),
+            "token": str(self.token),
+            "created_at": self.created,
+            "expires_at": self.expires_at,
+            "email": self.email,
+            "full_name": self.full_name,
+            "message": self.message,
+            "record_pid": self.record_pid,
+        }
+
+    @property
+    def need(self):
+        """Get the needs provided by the link."""
+        return AccessRequestTokenNeed(str(self.id))
+
+    @property
+    def is_expired(self):
+        """Determine if link is expired."""
+        return datetime.utcnow() > self.expires_at
+
+    @classmethod
+    def get_by_token(cls, token):
+        """Get the AccessRequestToken referenced by the specified token."""
+        return cls.query.filter_by(token=token).one_or_none()
+
+    @classmethod
+    def create(
+        cls,
+        email,
+        full_name,
+        message,
+        record_pid,
+        shelf_life=None,
+        expires_at=None,
+    ):
+        """Create a new request access token.
+
+        The ``expires_at`` value will take precedence over the ``shelf_life``.
+
+        :param email: The email address for which the request is made.
+        :param full_name: The full name of the requester.
+        :param message: The message attached to the request.
+        :param record_pid: The PID value for the record that is subject to the request.
+        :param shelf_life: The ``timedelta`` for how long the token will live for.
+        :param expires_at: The expiration date of the access request token.
+        """
+        if shelf_life and not expires_at:
+            expires_at = datetime.utcnow() + shelf_life
+
+        # TODO this was copy-pasted, check
+        is_date = isinstance(expires_at, date)
+        is_datetime = isinstance(expires_at, datetime)
+        if is_date and not is_datetime:
+            expires_at = datetime.combine(expires_at, datetime.min.time())
+
+        access_request_token = cls(
+            email=email,
+            full_name=full_name,
+            message=message,
+            record_pid=record_pid,
+            expires_at=expires_at,
+            token=token_urlsafe(
+                current_app.config.get("RDM_RECORDS_GUEST_ACCESS_TOKEN_LENGTH")
```

Comment:
probably we should factor in the `email` address and a `nonce` in the token generation rather than just generating a purely random token.
if we can extract these values from the (encrypted) token again, then we could use that to create more meaningful `Need`s from the token (like a `TokenNeed(nonce)`, and an `EmailNeed(email_address)`, which could help us with a few issues (like generating more sensible "comment author name" values on guest access requests).